scripts/check_compliance.py

```diff
@@ -426,73 +426,123 @@ class Codeowners(ComplianceTest):
+        # recursively glob the output of ls-files directly, only real
+        # files :-(
+
+        pattern2files = collections.OrderedDict()
+        top_path = Path(git_top)
+        top_len = len(git_top) + 1 # 1 for the slash
+
         with open(codeowners, "r") as codeo:
-            for line in codeo.readlines():
-                if not line.startswith("#") and line != "\n":
-                    match = re.match(r"([^\s]+)\s+(.*)", line)
-                    if match:
-                        add_base = False
-                        path = match.group(1)
-                        if path.startswith("/"):
-                            abs_path = git_root + path
-                        else:
-                            abs_path = "**/{}".format(path)
-                            add_base = True
-
-                        if abs_path.endswith("/"):
-                            abs_path = abs_path + "**"
-                        elif os.path.isdir(abs_path):
-                            self.add_failure("Expected / after directory '{}' "
-                                             "in CODEOWNERS".format(path))
-                            continue
-                        g = glob.glob(abs_path, recursive=True)
-                        if not g:
-                            self.add_failure("Path '{}' not found, in "
-                                             "CODEOWNERS".format(path))
-                        else:
-                            files = []
-                            if not add_base:
-                                for f in g:
-                                    l = f.replace(git_root + "/", "")
-                                    files.append(l)
-                            else:
-                                files = g
-
-                            all_files += files
-
-        files = []
-        for f in all_files:
-            if os.path.isfile(f):
-                files.append(f)
-
-        return set(files)
+            linenum = 0
+            for line in codeo:
+                linenum += 1
+                if line.startswith("#") or not line.strip():
+                    continue
+
+                match = re.match(r"^([^\s]+)\s+[^\s]+", line)
+                if not match:
+                    self.add_failure(
+                        "Invalid CODEOWNERS line %d\n\t%s" %
+                        (linenum, line))
+                    continue
+
+                git_patrn = match.group(1)
+                glob = self.git_pattern_to_glob(git_patrn)
+                files = []
+                for p in top_path.glob(glob):
+                    # comparing strings is much faster later
+                    files.append(str(p)[top_len:])
+
+                if not files:
+                    self.add_failure("Path '{}' not found, in "
+                                     "CODEOWNERS".format(git_patrn))
+
+                pattern2files[git_patrn] = files
+
+        return pattern2files
+
+
```

Comment:
Only one blank line between methods usually (though it's a bit hypocritical for me to complain, because I used to put a single blank line between everything... just to make it consistent :P).

Reply:
I admittedly didn't run pylint/pep8 yet, will do.